### ISSUE TITLE
Improved Binary Search (API Change Proposal)

### ISSUE THREAD

**Issue by liborty** [rust-lang/libs-team#81]
# Improved Binary Search

## The problem

Quoting from: [primitive.slice](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search):
"If there are multiple matches, then any one of the matches could be returned. The index is chosen deterministically, but is subject to change in future versions of Rust. If the value is not found then [Result::Err](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Err) is returned, containing the index where a matching element could be inserted while maintaining sorted order."

Some of these issues are since to some extent addressed by [partition_point](https://doc.rust-lang.org/std/primitive.slice.html#method.partition_point) with a different functionality. 

## Motivation, use-cases

* A typical application is efficient removal of duplicates of a particular value. For that, we need to find all (partially) equal items.
* In probability/statistics, we often want to count the number of occurrences of a particular event/value, intermixed amongst many other values/measurements within a sample. Then the solution is to sort the sample and then deploy my improved binary search, where range.end-range.start gives the answer. We can then keep the sorted list and possibly test for some other events/values in the same way. 
* Suppose you have a matrix/spreadsheet/database and instead of the usual retrieval of a value for some given coordinate(s), you need to solve the inverse problem, that is to find all the (pairs of) coordinates of where a particular value occurs. 

## Solution

The proposed solution has the following new benefits:

* It only requires PartialOrd instead of Ord.
* It works on both ascending and descending orders.
* It only searches the specified input Range<T>. This will often be quicker.
* The data can come from any source in address (index) space of type T and is only sampled (lazily) at the binary search points.
* The existing std `binary_search` varieties  use Err mechanism to deliver a genuine sort-order result, which is arguably not quite right. This is corrected. No errors are returned.
*  It solves the ambiguity about which (of possibly several)  match positions is returned. The result is now unambiguous: output range from the sort position of the first occurrence of the sought item till the last. Empty range n..n, means that the item was not found and n is its insert position. This mechanism is logically consistent and avoids doing Error processing of useful results.

## Links and related work

I have implemented the proposal as function `binary_find`.  It can be presently seen in my crate  [indxvec](https://github.com/liborty/indxvec/blob/main/src/lib.rs/#L66).

**Comment by the8472** [comment]
> A typical application is efficient removal of duplicates. For that, we need to find all (partially) equal items, not just one random equal item.

Can yo provide more details? Like how that isn't covered by sorting (which has to be done anyway for a binary search) and then using `Vec::dedup`?

> It works on both ascending and descending orders,

I don't see how this is necessary for deduplication.

> It solves the ambiguity hinted at in the above quote from std documentation.

But so does partition_point?

**Comment by liborty** [comment]
Thanks for your perceptive questions.

Yes, it is true that Vec::dedup will, in principle, solve the outlined example application. In practice, my proposal will often make it more efficient. For example, when we want to remove only duplicates of a specific value (possibly from lots of lists). Then it is wasteful to search linearly all of them for all the duplicates. We might not even want to remove the other duplicates at all.
Sorry, I should have written 'duplicates of some value' instead of just  'duplicates', implying all duplicates of all values, which is a different thing. I hear you say: "ok, so I pass a closure to `dedup_with` that tests for equality of only that value." Yes but then, to my knowledge, it will not deploy binary search.

Different orders may not be necessary for deduplication but when you already have a sort of a very long list, you do not want to have to  reverse it just because your binary search can cope with only one direction. Also, there are many other applications for a properly implemented general binary search.

Yes, I acknowledge that the `partition_point` also solves the Err returns but, then again, it has somewhat different functionality. I actually started by looking at `binary_search`, since it is there and I found it inadequate for my needs. Initially, I did not even suspect that there might be another method with an unrelated name which solves similar problem somewhat better. 
Yes, I should have studied documentation more thoroughly but I offer this as an argument for implementing `binary_search` properly, precisely because it is there and that is what people look at first.

**Comment by the8472** [comment]
> but when you already have a sort of a very long list, you do not want to have to reverse it just because your binary search can cope with only one direction. 

Or you can use `Reverse`:

```
    let slice = &[100usize, 80, 77, 4, 1];
    assert_eq!(Ok(1), slice.binary_search_by_key(&Reverse(&80), |e| Reverse(e)));
```

**Comment by liborty** [comment]
Another application:
In probability/statistics, we often want to count the number of occurrences of a particular event/value, intermixed amongst many other values/measurements within a sample. Then the solution is to sort the sample and then deploy my improved binary search, where range.end-range.start gives the answer. We can then keep the sorted list and possibly test for some other events/values in the same way. Not necessarily all of them.

**Comment by liborty** [comment]
> > but when you already have a sort of a very long list, you do not want to have to reverse it just because your binary search can cope with only one direction.
> 
> Or you can use `Reverse`:
> 
> ```
>     let slice = &[100usize, 80, 77, 4, 1];
>     assert_eq!(Ok(1), slice.binary_search_by_key(&Reverse(&80), |e| Reverse(e)));
> ```

Yes, or just use a complement index slice.len()-idx but when you have several occurrences, it will turn what was the first occurrence into the last and vice-versa

**Comment by liborty** [comment]
Here is a good 'killer application'.
Suppose you have a matrix/spreadsheet/database and instead of the usual retrieval of a value for some given coordinate(s), you need to solve the inverse problem, that is to find all the (pairs of) coordinates of where a particular value occurs.
* list all the values (you already have that in the scanning 1d sequence form of your matrix).
* list all their coordinates in associated lists, in corresponding positions.
* index sort the values
* use the sort index and my `binsearch_indexed` to efficiently look up all the coordinates for a particular value.

**Comment by scottmcm** [comment]
I'll cc https://github.com/rust-lang/rfcs/issues/2184, where something like this has been discussed (inspired by https://en.cppreference.com/w/cpp/algorithm/equal_range) for almost 5 years now.

I think the `-> Range<usize>` version of binary search makes sense to offer.  I don't know how to spell its name, though.

**Comment by liborty** [comment]
Yes, if the existing `binary_search` is to remain, then we need a new short and distinguished name. How about: `binary_find` ?

As to the functionality, there is an argument for keeping that simple too, as per my suggestion. That is, using implied PartialOrd instead of passing in a full blown comparator closure. After all, as has been rightly pointed out, the `partition_point` can be bent to that purpose. 

PartialOrd is already an improvement in generality on the Ord of the existing `binary_search`. 

Regards the above discussion about working on both ascending and descending orders: yes, that is to some extent optional but I have included it because it is practically a zero cost exercise ( a single test of a bool ).

**Comment by scottmcm** [comment]
I'm not on libs-api, but I'd be very surprised if a `bool` like that for the different orders got in.  

`.binary_search_range(x, true)` is neither readable nor rememberable, and while `.binary_search_range(x, Order::Ascending)` is readable, it's also super-verbose and the extra `enum` doesn't really pull its weight.

See also how there's no `bool` on `sort` for which way you want it sorted.

---

What you might consider instead -- as a separate ACP! -- would be something to make it easier to use `cmp::Reverse` to handle this situation.

For example, if there was a method to go `&[T] -> &[cmp::Reverse<T>]`, then you'd have things like
```
[5, 4, 2, 1].bikeshed_this().binary_search(Reverse(3)) // Err(2)
[1, 4, 2, 3].bikeshed_this().sort() // [4, 3, 2, 1]
```

**Comment by liborty** [comment]
See my above edit, probably before you read it.
It can be just: `.binary_find(x)` with order detected automatically

**Comment by liborty** [comment]
I have now implemented the automatic sort order detection, see 
[The Source](https://docs.rs/indxvec/1.3.3/src/indxvec/vecops.rs.html#126-224)

**Comment by liborty** [comment]
1. Your suggestion can be used to unite my proposal, let us call it `.binary_find()`, with the existing `binary_search()`.
2. We need to add a 'none' case:
```rust
a_slice.binary_finder().first() == a_slice.binary_find().start
a_slice.binary_finder().last() == a_slice.binary_find().end
a_slice.binary_finder().any() == a_slice.binary_search().ok()
a_slice.binary_finder().none() == a_slice.binary_search().err()
a_slice.binary_finder().range() == a_slice.binary_find()
```

**Comment by programmerjake** [comment]
imho it'd be useful to be able to binary search on non-slice things, e.g. a custom data type that is indexable, but data is not contiguous in memory.

**Comment by liborty** [comment]
Yes, exactly, that is why I am offering also `binsearch_indexed`

**Comment by liborty** [comment]
It works like this (self is a slice of any type &[T]):
```rust
    /// Binary search of an index sorted slice in ascending or descending order. 
    /// Like `binsearch` but using indirection via sort index `idx`.
    fn binsearch_indexed(self, idx:&[usize], val:&T) -> Range<usize> 
        where T: PartialOrd;
```
Where idx is obtained by:
```rust
    /// Makes a sort index for self, using key generating closure `keyfn`
    fn keyindex(self, keyfn:fn(&T)->f64, ascending:bool) -> Vec<usize>;
```
The beauty of it is that T can be any arbitrarily complex user defined type but as long as you supply the `keyfn` (closure), the index is built using my super-fast `hashsort` (over f64 keys). I could subsume `keyindex` within `binarysearch_indexed` but I find that having the explicit index available is often useful for other purposes as well.

When calling `keyindex`, the closure can capture any additional variables it might need. I use it in `rstats` crate to sort whole vectors, with keys derived as any (scalar) function on the vectors, such as a their distance from some point.

**Comment by programmerjake** [comment]
> It works like this (self is a slice of any type &[T]):
> 
> ```rust
>     /// Binary search of an index sorted slice in ascending or descending order. 
>     /// Like `binsearch` but using indirection via sort index `idx`.
>     fn binsearch_indexed(self, idx:&[usize], val:&T) -> Range<usize> 
>         where T: PartialOrd;
> ```

that's nice, but it doesn't actually do what imho is needed, which is more like:
```rust
// TODO: pick better name
// TODO: add versions with u32, u64, and u128 instead of usize -- needed
// because you might be searching on disk or over the network rather
// than in-memory and usize can be too small.
// TODO: switch to using Try trait rather than Result.

// Note how this needs no slice to be constructed in order to
// search, not everything that can be searched fits in memory or
// wants to pay the up-front cost of building a slice, e.g. if you're
// searching in something with 400 trillion entries, nearly no one
// has that much memory to construct a slice of, nor wants to
// spend hours filling memory with indexes before it can conduct
// a simple <1s search.
pub fn try_bin_search_algo<E>(mut search_in: Range<usize>, mut cmp_at: impl FnMut(usize) -> Result<Ordering, E>) -> Result<Result<usize, usize>, E> {
    // hope i got this right
    while !search_in.is_empty() {
        let mid = (search_in.end - search_in.start) / 2 + search_in.start;
        match cmp_at(mid)? {
            Ordering::Less => search_in.end = mid,
            Ordering::Greater => search_in.start = mid + 1,
            Ordering::Equal => return Ok(Ok(mid)),
        }
    }
    Ok(Err(search_in.start))
}
```

**Comment by liborty** [comment]
So, you envisage combining this with a random disk/internet access?  I have not thought that far.

**Comment by programmerjake** [comment]
> So, you envisage combining this with a random disk/internet access?

yes, or something else other than just an index-into-slice-and-compare, e.g.:
```rust
// compute sqrt(a) via binary search.
try_bin_search_algo(0..(1 << 32), |v| Ok(if Some(v2) = v.checked_mul(v) { a.cmp(&v2) } else { Ordering::Less }))
```

> I have not thought that far. Does that not warrant a separate method?

probably...

**Comment by liborty** [comment]
Yes, like I said, I can subsume the `keyindex` method and just pass the closure to the binary search directly.
You are quite right that doing only the comparisons with a closure probe 'on demand', saves time and space for generating all the key values. Thank you for that!

**Comment by cuviper** [comment]
It seems to me that there is enough design space left that this isn't ready for the standard library yet.

**Comment by programmerjake** [comment]
This makes me think that Rust needs something in the standard library like C++'s `<algorithms>` header, where we can put binary search, sorting algorithms (which also need a version that doesn't need a slice to sort), and more.

Maybe `core::algo::*`?

**Comment by liborty** [comment]
I have taken all your suggestions into account and incorporated them into this definitive proposal. Take it or leave it.
In any case, it is available in my  `indxvec` crate.
 
I went for a good compromise between generality and ease of use.
Towards the latter objective, I decided against passing in a comparator closure, as the same effect can be achieved, imho in a cleaner way, by implementing `PartialOrd` for your custom data type T.
Nonetheless, I am passing in a `probe` closure to enable the flexible data sources, including monotonic functions.

I stayed with usize for the indexing type but I give an example transformation function to f64, used in the tests below by the example general roots finder. I hasten to add that this is not the best way of solving equations but as long as the root is bracketed by the given range, it is guaranteed to converge. If it is outside the given range, `binary_find` will report an error and the direction in which the range needs to be extended.

```rust
/// transforms i:usize which is normally in the full usize range
/// to f64, such that roots of num are bracketed
fn uasf(num:f64,i:usize) -> f64 {
    num*(i as f64)/(usize::MAX as f64)
}
```

```rust
/// General Binary Search
/// Search only within the specified range, which is always ascending. 
/// Probe obtains the data value T at key:usize, by any means, 
/// from any kind of sorted data or monotonic function.
/// The sort order can be either ascending or descending (increasing/decreasing).
/// PartialOrd has to be implemented separately for custom types T.
/// When item is in order before range.start, Err(range.start) is returned.
/// When item is in order after range.end, Err(range.end) is returned.
/// Otherwise binary_find returns Range of all the consecutive values PartiallyEqual to the sought item:&T.
/// When item was not found, then the returned_range will be empty and 
/// returned_range.start (and end) will give the sort position where the item can be inserted.
pub fn binary_find<T,F>(range:Range<usize>,probe: F, item:&T )  
    -> Result<Range<usize>,usize> where T:PartialOrd, F:Fn(usize)-> T { 

    let last = |idx:usize| -> usize {
        let mut lastidx = idx+1;
        for i in idx+1..range.end { // move end up
            if item == &probe(i) { lastidx += 1; } else { break; }; 
        }
        lastidx
    };
    let first = |idx:usize| -> usize {
        let mut firstidx = idx;
        for i in (range.start..idx).rev() { // move start down
            if item == &probe(i) { firstidx -= 1; } else { break; }; 
        }
        firstidx
    }; 
 
    let firstval = probe(range.start); 
    let lastval = probe(range.end-1);
    // when data is in descending order, reverse all comparisons
    let ordered = if firstval < lastval { |a:&T,b:&T| a < b } 
    else { |a:&T,b:&T| b < a }; // comparisons closure defined by the sort order 

    if ordered(item,&firstval) { return Err(range.start); } // item is before the range.start
    else if ordered(&lastval,item) { return Err(range.end); } // item is beyond the range.end
    else if firstval == lastval { return Ok(range); } // range data is all equal to item or empty 

    if item == &firstval { // item is equal to the first data item
        return Ok(range.start..last(range.start));
    };
    if item == &lastval { // item is equal to the last data item in range
        return Ok(first(range.end-1)..range.end);
    };
    let mut hi = range.end - 1; // initial high index
    let mut lo = range.start; // initial low index
    loop {
        let mid = lo + (hi-lo) / 2; // binary chop here with truncation
        if mid > lo { // still some range left
            let midval = probe(mid);
            if ordered(&midval,item) { lo = mid; continue; };
            if ordered(item,&midval) { hi = mid; continue; }; 
            // neither greater nor smaller, hence we found match(es) 
            return Ok(first(mid)..last(mid));            
        }
        else { return Ok(hi..hi) }; // interval is exhausted, val not found
    }
}
```
The test source (examples of usage and functionality):

```rust
#[test]
fn text() {
    let sentence = 
        "Humpty Dumpty sat on a wall, \
        Humpty Dumpty had a great fall, \
        and all the king's horses and all the king's men \
        could not put Humpty together again";
    let v = sentence.split(' ').collect::<Vec<_>>();
    println!("{}",v.gr()); // Display
    // using cloning mergesort with implied alphabetic partial order 
    let mut sorted = v.sortm(true);
    println!("Ascending sorted:\n{}",sorted.gr());
    // Binary search using implied alphabetic partial order
    println!("Binary_search for {BL}'Humpty'{UN}: {YL}{:?}{UN}",
        binary_find(0..sorted.len(),|i|sorted[i], &"Humpty"));
    println!("Binary_search for {BL}'Humpty'{UN} in range 5..end: {YL}{:?}{UN}", 
        binary_find(5..sorted.len(),|i|sorted[i], &"Humpty"));
    println!("Binary_search for {BL}'the'{UN}: {YL}{:?}{UN}", 
        binary_find(0..sorted.len(),|i|sorted[i], &"the"));
    println!("Binary_search for {BL}'the'{UN} in range 0..24: {YL}{:?}{UN}", 
        binary_find(0..24,|i|sorted[i], &"the"));
    println!("Binary_search for {BL}'queen's'{UN}: {YL}{:?}{UN}", 
        binary_find(0..sorted.len(),|i|sorted[i], &"queen's"));
    sorted.dedup();
    println!("Dedup:\n{}\n",sorted.gr());    
    let mut dsorted = v.sortm(false);
    println!("Sortm descending sorted:\n{}",dsorted.gr());
    // Cloning binary search using implied alphabetic partial order
    println!("Binary_search for {BL}'Humpty'{UN}: {YL}{:?}{UN}",
        binary_find(0..dsorted.len(),|i|dsorted[i], &"Humpty"));
    println!("Binary_search for {BL}'Humpty'{UN} in range 0..22: {YL}{:?}{UN}", 
        binary_find(0..22,|i|dsorted[i], &"Humpty"));
    println!("Binary_search for {BL}'the'{UN}: {YL}{:?}{UN}", 
        binary_find(0..dsorted.len(),|i|dsorted[i], &"the"));
    println!("Binary_search for {BL}'the'{UN} in range 5..end: {YL}{:?}{UN}", 
        binary_find(5..dsorted.len(),|i|dsorted[i], &"the"));
    println!("Binary_search for {BL}'queen's'{UN}: {YL}{:?}{UN}", 
        binary_find(0..dsorted.len(),|i|dsorted[i], &"queen's"));
    dsorted.dedup();
    println!("Dedup:\n{}\n",sorted.gr());

    let num = 5.;
    let root = 3;
    // find 1/nth root of num
    if let Ok(r) = binary_find(0..usize::MAX,
        |i| { uasf(num,i).powi(root) },&num)  
        { println!("1/{} power of {} {} {}", 
          root.yl(), num.yl(), num.powf(1./root as f64).gr(),uasf(num,r.start).gr()); };
}
```
And its output:
```rust
running 1 test
[Humpty Dumpty sat on a wall, Humpty Dumpty had a great fall, and all the king's horses and all the king's men could not put Humpty together again]
Ascending sorted:
[Dumpty Dumpty Humpty Humpty Humpty a a again all all and and could fall, great had horses king's king's men not on put sat the the together wall,]
Binary_search for 'Humpty': Ok(2..5)
Binary_search for 'Humpty' in range 5..end: Err(5)
Binary_search for 'the': Ok(24..26)
Binary_search for 'the' in range 0..24: Err(24)
Binary_search for 'queen's': Ok(23..23)
Dedup:
[Dumpty Humpty a again all and could fall, great had horses king's men not on put sat the together wall,]

Sortm descending sorted:
[wall, together the the sat put on not men king's king's horses had great fall, could and and all all again a a Humpty Humpty Humpty Dumpty Dumpty]
Binary_search for 'Humpty': Ok(23..26)
Binary_search for 'Humpty' in range 0..22: Err(22)
Binary_search for 'the': Ok(2..4)
Binary_search for 'the' in range 5..end: Err(5)
Binary_search for 'queen's': Ok(5..5)
Dedup:
[Dumpty Humpty a again all and could fall, great had horses king's men not on put sat the together wall,]

1/3 power of 5 1.7099759466766968 1.709975946676697
test text ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.00s
```

**Comment by CAD97** [comment]
> I decided against passing in a comparator closure, as the same effect can be achieved, imho in a cleaner way, by implementing `PartialOrd` for your custom data type `T`

Just to note, this isn't always possible. E.g., in string interners, it's really nice to store

```rust
type Span = Range<u32>;
type Symbol = NonZeroU32;
pub struct Interner<S> {
    hasher: S,
    string_to_symbol: HashSet<Symbol, ()>,
    symbol_to_span: Vec<Span>,
    span_to_string: String,
}
```

with then hash/comparison being of `&str <=> &span_to_string[symbol_to_span[symbol]]`. If the key actually stored in the data structure you're searching is a key and not an index, it cannot implement `Partial`/Ord`/`Eq` directly; it needs to store an actual reference back to the shared resource. And in a case like this, that's not even really possible, since the `String` can be appended to and reallocate, invalidating extant references.

This is just what I'm the most familiar with, but this pattern of `raw_entry` usage to do "dependency injection" for comparisons is a surprisingly commonly available optimization once you know how to look for it.

**Comment by liborty** [comment]
I struggle to understand this. 
Given that you are being asked to implement PartailOrd on your own custom type T,
how is it not possible attach the same trics to T?

**Comment by CAD97** [comment]
See https://github.com/CAD97/strena as an example. The HashMap `raw_entry` API allows you to provide dependency injected `fn hash(&K) -> u64` and `fn eq(&Q, &K)` for lookup.

**Comment by liborty** [comment]
Then why not just do your binary search over the symbols?

**Comment by CAD97** [comment]
Because the symbols aren't sorted intrinsically. They're indexes, sorted by the string which they reference, which to recover you need to provide outside data. [ex.:](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f00a0a4d186fd6400d995ca86fd637be)

```rust
pub type Symbol = usize;
pub type Span = Range<usize>;

pub struct Interner {
    string_to_symbol: Vec<Symbol>,
    symbol_to_span: Vec<Span>,
    span_to_string: String,
}

impl Interner {
    pub fn get(&self, s: &str) -> Option<Symbol> {
        let string_to_symbol: &[Symbol] = &self.string_to_symbol;
        let symbol_to_span: &[Span] = &self.symbol_to_span;
        let span_to_string: &str = &self.span_to_string;

        let range = binary_find(0..string_to_symbol.len(), |ix| {
            s.cmp(&span_to_string[symbol_to_span[string_to_symbol[ix]].clone()])
        });
        Some(range)
            .filter(|range| !range.is_empty())
            .map(|range| string_to_symbol[range.start])
    }
}
```

**Comment by liborty** [comment]
Yes but now every  cmp closure has to check for the descending order on every invocation, which is not practical. I you can solve that problem, I will be happy to adopt your proposal.

Also, we can not return an empty range when the searchee is outside it, as the empty range is reserved for indicating the insert order. However, this is not such a problem and if you consult my latest code, you will see that I have solved it already by returning the out of range index as error.

I like the probe closure being explicit. I think it makes the code much easier to follow than burying it inside the cmp closure. Any reason why we can not have both probe and cmp?

**Comment by CAD97** [comment]
> Yes but now every cmp closure has to check for the descending order on every invocation

No, because you just prestore that information in the closure the same way you're doing currently. Or use `cmp::Reverse`, or whatever; Rust's sorting APIs do not assume which order you want, they require you to put in the correct ordering.

If anything, this is better, because with this version you can pass in a static order whereas your code always does a dynamic comparison ordering.

> Also, we can not return an empty range when the searchee is outside it, as the empty range is reserved for indicating the insert order of a missing searchee within the range.

If I return `begin..begin`, that means that the insertion position is before the first element, right? Same for `end..end` is after the end element. If you want to know if the element is strictly contained in the range, this isn't the API that's serving that purpose.

Though that does mean the `end..end` return at the end is wrong. (I mostly just copied your impl, assuming it good, fixing up types.) It should be `mid..mid`, where `mid` is the insertion point to retain a sorted order.

> I like the probe closure being explicit. I think it makes the code much easier to follow than burying it, the searchee and everything inside the opaque cmp captured environment. This seems just as bad as any global variables ever were. Any reason why we can not have both probe and cmp?

I'm going to turn this question around. What benefit is there to having a separate index and compare step? Having it in one step saves a (potentially) dynamic call. Rust likes its stateful closures; in fact, this should probably even take `FnMut`.

Having a unified compare does have benefits, in that it (potentially) lowers monomorphization pressure, but more importantly, your API requires the key to be sized, copy, `'static` (because you extract the key by value `fn(usize) -> T`) whereas using a single `cmp` closure allows the indexed values to be ephemeral to the comparison closure, by reference or by value.

-----

Also, always indexing in `u128` is going to be a big problem. `u128` is emulated and slow on plenty of platforms, and you physically can't have more than `isize::MAX` district items anyway.

We probably need to get the bounds by continuing the binary search as well, rather than falling back to a linear probe. Or at a minimum use some sort of exponential probe to avoid the worst case performance of `[0, [1; BIG]..., 2]`.

**Comment by liborty** [comment]
> If I return `begin..begin`, that means that the insertion position is before the first element, right? Same for `end..end` is after the end element. If you want to know if the element is strictly contained in the range, this isn't the API that's serving that purpose.

Actually, `begin..begin` means that the insertion position is at the first element.

> Also, always indexing in u128 is going to be a big problem. u128 is emulated and slow on plenty of platforms, and you physically can't have more than isize::MAX district items anyway.

I was just trying to accommodate [programmerjake](https://github.com/programmerjake) and his request to widen the address  space to disks, internet, etc. It is certainly an intriguing idea to binary search the internet.

**Comment by CAD97** [comment]
> Actually, `begin..begin` means that the insertion position is at the first element.

AIUI, given a sorted two-element slice `[ a , b ]` and binary searching for `x`, the correct semantics are

```
0..0 => [ x , a , b ].is_sorted()
1..1 => [ a , x , b ].is_sorted()
2..2 => [ a , b , x ].is_sorted()
0..1 => x == a && x < b
1..2 => a < x && x == b
0..2 => x == a && x == b
```

and these are all possible outcomes of a binary search on a sorted list. Where do we disagree?

I *suppose*, if this is a subslice, insertion at the edges may not maintain the entire slice being sorted. However, this is still a determinable output condition, and there's no reason to assume that binary searching a subslice that your item is not in will give you a position that keeps the superslice still sorted; the correct answer is to say inserting at the edge would keep the input slice range sorted.

I think the more important distinction, though, is that "insertion *at* an element" is *meaningless*. You insert *between elements*. The range `0..0` is an empty range which is *before* the first element, as is observed with [`Vec::splice`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.splice).

When I say "insertion position is before the first element," I do mean the position directly before the first element.

----

I'm also going to repeat my counterquery for visibility:

> What *benefit* is there to having a separate index and compare step?

Because the benefit of *not* is that a separate index step really wants a signature of `fn index(&mut self, ix: usize) -> Self::Key<'_>`; that is, return a key which can *optionally* borrow from the container and only needs to be valid until the next call to `index`. (Think "random access lending iterator".)

**Comment by liborty** [comment]
> 0..0 => [ x , a , b ].is_sorted()
and these are all possible outcomes of a binary search on a sorted list. Where do we disagree?

Ah, I see what is going on! I was referring to the subscripts of the input range, as supplied. Assuming that any insertion process would on seing `0..0` insert x  in position 0 and shift the rest one up. Whereas you are assuming that this has already been done and your subscripts are referring to the new slice. So, we do not disagree.

On the count of having an all-in-one comparator, I think you persuade me about the benefits.
 
I accept that the current functions/methods implicitly assume that everything is in ascending order but that seems to me an unwarranted assumption. In any case, I would like to relax it. I think that having both orders automatically recognised and correctly acted upon has benefits in generality and code robustness. Presumably, this will now have to be done externally to `binary_find`. Where an if statement similar to mine will be issuing two different calls on the binary_find, with two different codes for the cmp closure, one involving Reverse? Though, as you say, not everyone will necessary need that.

Thank you for your clarifications, I find them very useful.

**Comment by CAD97** [comment]
> Presumably, this will now have to be done externally, issuing two different calls on the binary_find

No, one call is sufficient; just do a `let cmp = if ascending { |x| key.cmp(x) } else { |x| key.cmp(x).reverse() }` like indxvec currently does internally, or `let cmp = |x| { let c = key.cmp(x); if ascending { c } else { c.reverse() } }`. (The latter might have better inlining behavior.)

> I accept that the current functions/methods implicitly assume that everything is in ascending order but that seems to me an unwarranted assumption.

For intrinsic ordering, I *might* agree with you. But when injecting a comparator/ordering, it is absolutely reasonable to assume that the ordering is the same that the ordered input is ordered by.

**Comment by liborty** [comment]
> ... no? Given `index..index`, I should do `vec.insert(index, item)` to insert it and maintain the sort order. Yes, that results in `vec[index]` being `item`, but this is because we inserted it at that index. The index is _both_ the insertion index and the index after insertion, because that's just restating the same thing twice.

Yes, of course. But when you said `0..0` is before the range and I said that a is at 0, you were thinking "as if the outside item x was inserted there" whereas I was thinking of the original range as given. Anyway, it does not matter, I understand what you mean. I can certainly adopt this  and get rid of the Err, never liked it anyway :)

**Comment by liborty** [comment]
Thanks to your excellent suggestions we are making progress. I have now implemented all of them apart from replacing also the linear search of matching items with binary search. In a few extreme cases it will be much quicker but also, in many usual cases where there are only a few matching items, it will be slower.  If there is a feeling that it is worth it, I can certainly do that: basically a recursive call with a comparator that looks at two adjacent items to find the edge cases.

The biggest latest improvement  I made is generic search Range<T>, so that the user can choose what index type suits best: `usize`, `u128`, or anything else (satisfying the generic trait bounds). Here is an example application to root finding, using f64:

```rust
use core::ops::Range;
#[test]
fn roottest() { 
    let num:f64 = 1234567890.0;
    let root:f64 = 5.3;
    let range = broot(num,root);
    println!("{} to the power of {YL}1/{}{UN}\nbinary_find:\t{} error: {}\npowf:\t\t{} error: {}", 
    num.yl(), root, range.start.gr(), (num-range.start.powf(5.3)).gr(),
    num.powf(1./root).gr(),(num-num.powf(1./root).powf(root)).gr()); 
}

///num.powf(1./root) using binary search
fn broot(num:f64,root:f64) -> Range<f64> {
    binary_find(1_f64..num,
        |&probe| { 
            if probe.powf(root) < num { Less }
            else if probe.powf(root) > num { Greater }
            else { Equal }})
}
running 1 test
1234567890 to the power of 1/5.3
binary_find:    51.92543988913025 error: -0.000000476837158203125
powf:           51.92543988913026 error: -0.0000011920928955078125
test roottest ... ok
```

**Comment by CAD97** [comment]
In std, the range index type would probably be bound by [`Step`](https://doc.rust-lang.org/nightly/core/iter/trait.Step.html)[^1]. This can theoretically even be accomplished exclusively with `step_unchecked` and without cloning the range keys, though I don't know what the performance impact of doing so would be; std would probably want to specialize on `Copy` to use the simple `let mid = step_halfway(lo, hi)` approach.

[^1]: `Step` is surprisingly close to C++'s named requirement LegacyRandomAccessIterator / concept `random_access_iterator`, minus the dereference for item access of course. (Note C++ copy constructors are Rust `Clone`.)

However, I'm bowing out on design here. I have strong opinions on the proper dependency injection API, but less so on the rest of the details.

(Though I do want to note that in parallel to `raw_entry`, perhaps this should )

**Comment by programmerjake** [comment]
more than just `Step` would be needed to search in `0..!0u128`, since that exceeds the range of `usize` by a large factor. you'd probably want:
```rust
trait StepHalfway: Step {
    /// returns `(self + other) / 2` but without overflowing.
    fn halfway(&self, other: &Self) -> Self;
}
```

**Comment by programmerjake** [comment]
also it shouldn't be limited to just integer types, e.g. binary searching a list of SHA1 or SHA256 hashes would be very useful for git stuff.

**Comment by CAD97** [comment]
`Step` eventually should be able to be stabilized and implementable for any type that fits its interface.

**Comment by scottmcm** [comment]
As a meta-comment, this thread feels very much like what I'd expect in a *crate* design discussion.  That's not saying that these ideas are bad, but the super-general version of all this feels like something that belongs in a crate, not in `core`, since it's far from the plausible original "hey, can we just get one method on slice that's a bit better for these cases".

**Comment by liborty** [comment]
I must be missing something in this discussion of `step`: `binary_find` works just fine on u128 already, since several versions back. Also on f64 which is not `step` as you can see from my latest example above. This is achieved, in general, by directly implementing the comparator.

**Comment by programmerjake** [comment]
some simple reasons to have it be `FnMut`: reusing a http connection, or where you need to cache partial results.

**Comment by CAD97** [comment]
> I only ever need to add/subtract 1 and divide by 2. With this, I believe I am still within the scope of: "hey, can we just get one method on slice that's a bit better for these cases".
>
> This is the signature:
> 
> ```
> pub fn binary_find<K, V>(range: Range<T>, cmp: F) -> Range<T>
where
    T: PartialOrd + Copy + Add<Output=T> + Sub<Output=T> + Div<Output=T> + From::<u8>,
    F: Fn(&T) -> Ordering,
```

(Reformatted to be legible. *Please*, use rustfmt, at least if you're posting code on rust-lang repositories. Maintaining a consistent style helps code to be quickly understood.)

This signature is still *vastly* more complicated than any other bound in the standard library APIs. And it's not even what you say it is! You're not requiring "add one and divide by two", you're requiring `From<u8>` and `Div`! How would this make any sense for e.g. `Uuid`, `Sha1`, or other types which are `Step` (have predecessor/successor operations) but not just numbers.

`(x: Sha1) / Sha1::from(2)` is nonsensical.

`Step` (or ) *is* the semantically correct bound here, as random-access successor/predecessor is what you need to do a binary search. If `Step` isn't powerful enough, then it should be generalized and/or extended to support this usecase.

That said, though, binary searching over a dataset with more elements (which is sorted in some fashion!) than the address space is an extremely niche use case. And will likely want to switch strategies when moving from impossibly sparse to some reasonable scale to work on.

**Comment by liborty** [comment]
Your formatting has actually destroyed my signature snippet. It mismatched the trait arguments and chopped off the end of the bounds. No wonder you have difficulties reading it.

Are we talking about the same thing here? My type T is the type of the index, not of the data.
How do you propose to find a midpoint of an uncountable index range?

**Comment by timvermeulen** [comment]
I second the request to please use rustfmt in these discussions. I've had some (slight) difficulties reading these code snippets because I'm so used to reading Rust code that is formatted using it.

**Comment by liborty** [comment]
OK, here it is:
```rust
pub fn binary_find<T, F>(range: Range<T>, cmpr: F) -> Range<T>
where
    T: PartialOrd + Copy + Add<Output = T> + Sub<Output = T> + Div<Output = T> + From<u8>,
    F: Fn(&T) -> Ordering,
```

**Comment by timvermeulen** [comment]
I've seen exponential search used for this, too — still O(log n) in the worst case, but clearly better (or at least fewer comparisons) than binary search for a small number of equal elements.

**Comment by liborty** [comment]
Great advice, thanks! I will do that next.

**Comment by liborty** [comment]
I have just realised that I can get even better results by reusing the last bounds of the already performed binary search!

**Comment by CAD97** [comment]
> The bounds may look complicated but it is not my fault that there is no `Countable` trait in Rust.

*As I've been saying*, that trait is `Step`.

> How do you propose to find quickly a midpoint of an uncountable index range?

```rust
let distance = Step::steps_between(&start, &end)?;
let mid = Step::forward_unchecked(start, distance / 2);
```

With `Range<impl Step>`, it's *trivially* possible to bridge to working `Range<usize>` if you're okay with extra clones of the index type:

```rust
pub fn binary_find<K: Step>(range: Range<K>, cmp: impl Fn(&K) -> Ordering) -> Range<K> {
    let trans = |i: usize| Step::forward(range.start.clone()).unwrap();
    if let Some(length) = Step::distance_between(&range.start, &range.end) {
        let out = your_binary_find(0..length, |i| cmp(&trans(i));
        trans(out.start)..trans(out.end)
    } else {
        range
    }
}
```

(Sorry about the broken formatting in my last post I did it on a phone without checking the output and just before I went to sleep.)

**Comment by liborty** [comment]
That sounds good. Do you have any clearer idea, how it actually obtains the `distance_between`? Because if it just linearly counts the steps, that would be too slow. If not, then I will attempt to change to it next.

Here is my [current version](https://github.com/liborty/indxvec/blob/main/src/lib.rs#L64) which implements everything  up to here, except that `Step`. It now includes the binary search for the end(s) of the matching range, restricted to the confines of the final range of the binary search for that first match. Which is a great idea, even if I say so myself :)

**Comment by programmerjake** [comment]
> ```rust
> pub fn binary_find<K: Step>(range: Range<K>, cmp: impl Fn(&K) -> Ordering) -> Range<K> {
>     let trans = |i: usize| Step::forward(range.start.clone()).unwrap();
>     if let Some(length) = Step::distance_between(&range.start, &range.end) {
>         let out = your_binary_find(0..length, |i| cmp(&trans(i));
>         trans(out.start)..trans(out.end)
>     } else {
>         range
>     }
> }
> ```

so if you want to search in the range `0..!0u128` it just always returns the input range!? that's just plain broken imho. `Step` by itself currently doesn't have the functionality needed to efficiently binary search ranges larger than `usize`.

**Comment by CAD97** [comment]
I keep saying the same things.

- Yes, `Step` is unstable.
  - But that doesn't matter.
  - We're not discussing the implementation of a crate that works on stable.
  - We're discussing a new unstable API to potentially add to std.
- Yes, `Step` is limited to getting a max distance of `usize`.
  - But because `Step` is unstable, we can change that if desired.
    - Probably remove `steps_between` from `Step` and add it to a subtrait.
  - It's not necessarily needed, though, as you can binary search buckets and then binary search within the buckets.

e.g. smth like untested

```rust
fn split(x: i128) -> (i64, u64) {
    (((x as u128) >> 64) as u64 as i64, x as u128 as u64)
}
fn unsplit(hi: i64, lo: u64) -> i128 {
    lo as i128 & ((hi as u64 as u128) << 64) as i128
}

fn search(range: Range<i128>, cmp: impl Fn(i128) -> Ordering) -> Range<i128> {
    let cmp_hi = |&hi| cmp(unsplit(hi, 0));
    let hi = binary_find(split(range.start).0..=split(range.end).0);
    let cmp_lo = |hi| |&lo| cmp(unsplit(hi, lo));
    let lo = binary_find(0..=u64::MAX, cmp_lo(hi.start)).start..binary_find(0..=u64::MAX, cmp_lo(hi.end)).end;
    unsplit(hi.start, lo.start)..unsplit(hi.end, lo.end)
}
```

(requires `binary_find` support for `RangeInclusive` / finite `impl RangeBounds`, and I remain sad that ranges being `Iterator` instead of just `IntoIterator` means `RangeInclusive` doesn't expose fields.)

**Comment by programmerjake** [comment]
> ```rust
> fn split(x: i128) -> (i64, u64) {
>     (((x as u128) >> 64) as u64 as i64, x as u128 as u64)
> }
> fn unsplit(hi: i64, lo: u64) -> i128 {
>     lo as i128 & ((hi as u64 as u128) << 64) as i128
> }
> 
> fn search(range: Range<i128>, cmp: impl Fn(i128) -> Ordering) -> Range<i128> {
>     let cmp_hi = |&hi| cmp(unsplit(hi, 0));
>     let hi = binary_find(split(range.start).0..=split(range.end).0);
>     let cmp_lo = |hi| |&lo| cmp(unsplit(hi, lo));
>     let lo = binary_find(0..=u64::MAX, cmp_lo(hi.start)).start..binary_find(0..=u64::MAX, cmp_lo(hi.end)).end;
>     unsplit(hi.start, lo.start)..unsplit(hi.end, lo.end)
> }
> ```

that has several issues:
* it searches outside the input range, e.g. if start is `0x10000000001000000` it will search starting from `0x10000000000000000`. This is an issue because some comparison functions will fail if an index outside the given range is passed in, e.g. `|idx| (5000000000 / (idx - (1 << 64))).cmp(&0)` will divide by zero if called on `0x10000000000000000`
* it only works if `usize` is 64-bit. if it's 32 or 16-bit it'll need to be orders of magnitude more complex.

imho if I'd have to go through all that just to binary search on `u128`, i'd rather just write the binary search myself to work on `u128` -- defeating the entire purpose of having binary search in `core` in the first place.

Therefore, I think we need `fn halfway(&self, other: &Self) -> Self` in either `Step` or some other trait, since the current `Step` is totally insufficient.

**Comment by liborty** [comment]
> * it searches outside the input range

I will second that. I have gone into quite a lot of trouble to avoid any such access, even to `range.end`. If for no other reason than that will panic with 'out of range' error even on an ordinary Vec.

> imho if I'd have to go through all that just to binary search on u128, i'd rather just write the binary search myself to work on u128 -- defeating the entire purpose of having binary search in core in the first place.

Which is what I have done and proposed. The only benefit I see to Step is that it is somehow 'more idiomatic' but the complications are significant. I always took 'idiomatic' as equivalent to 'simpler' and/or 'more general', otherwise there is no point. My explicit trait bounds are more precise and fulfil the purpose right now, in stable Rust.

>Therefore, I think we need fn halfway(&self, other: &Self) -> Self in either Step or some other trait, since the current Step is totally insufficient.

Yes, I agree, this would persuade me. I note that it would have to be implemented internally in much the same way as what I had done already,  i.e. as a numerical division.

**Comment by CAD97** [comment]
It is clear that we are at this point no longer discussing an API for std but instead the implementation of indxvec. I'm muting this.

We already have [`slice::pub fn binary_search_by<'a, F>(&'a self, f: F) -> Result<usize, usize>`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.binary_search_by). Adding `slice::binary_find[_by] -> Range<usize>` seems reasonable. Indirected binary find that works just on `Range` (or perhaps a `RandomAccessIterator` like is already used internally in std) is *possible* but clearly not ready for std inclusion as there are too many open design questions.

(The implementation for i128 is *an example* of the tiered search technique, not a proposal of how to do it. It would be not too difficult to fix to avoid the mentioned deficiencies.)

**Comment by liborty** [comment]
I have now completed the general [design & implementation](https://github.com/liborty/indxvec), satisfying all the requirements mentioned in these discussions. I commend it to your attention and adoption. It is in the trait `Search`. You can read the description of my new algorithm in README.md.

It is true that it now applies to Range<T> rather than to a slice, thus formally justifying the closure of this proposal. However, I think you would do well to adopt it in some form or another.

It is not using Step, which is not ready and would not, in any case, enable solving equations with Range<f64>, as  my design does. Together with any other numerical  indices, such as usize, u128, etc.
